fix(wash-up): grant execute permission to `mac_listener` for hot-reloading

nihil2501 opened the pull request:
PR to demonstrate the essence of a fix for hot reloading actors on my system.
```
11:51:56.091 [info] Starting actor MDPJTMU43XXM5KWADFSEI7LM7IL3SSVU565DXIJXCH457TPPI2VU7WCT
11:51:56.211 [error] GenServer :actor_watcher terminating
** (MatchError) no match of right hand side value: :ignore
    (wasmcloud_host 0.57.1) lib/wasmcloud_host/actor_watcher.ex:91: WasmcloudHost.ActorWatcher.handle_call/3
    (stdlib 4.0.1) gen_server.erl:1146: :gen_server.try_handle_call/4
    (stdlib 4.0.1) gen_server.erl:1175: :gen_server.handle_msg/6
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message (from #PID<0.2272.0>): {:hotwatch_actor, "/path/to/myactor_s.wasm", 1}
11:51:56.212 [error] GenServer #PID<0.2272.0> terminating
** (stop) exited in: GenServer.call(:actor_watcher, {:hotwatch_actor, "/path/to/myactor_s.wasm", 1}, 60000)
    ** (EXIT) an exception was raised:
        ** (MatchError) no match of right hand side value: :ignore
            (wasmcloud_host 0.57.1) lib/wasmcloud_host/actor_watcher.ex:91: WasmcloudHost.ActorWatcher.handle_call/3
            (stdlib 4.0.1) gen_server.erl:1146: :gen_server.try_handle_call/4
            (stdlib 4.0.1) gen_server.erl:1175: :gen_server.handle_msg/6
            (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
    (elixir 1.13.4) lib/gen_server.ex:1030: GenServer.call/3
    (wasmcloud_host 0.57.1) lib/wasmcloud_host_web/live/components/start_actor_component.ex:69: StartActorComponent.handle_event/3
    (phoenix_live_view 0.16.4) lib/phoenix_live_view/channel.ex:553: anonymous fn/4 in Phoenix.LiveView.Channel.inner_component_handle_event/4
    (telemetry 0.4.3) /Users/brooks/github.com/wasmcloud/wasmcloud-otp/wasmcloud_host/deps/telemetry/src/telemetry.erl:272: :telemetry.span/3
    (phoenix_live_view 0.16.4) lib/phoenix_live_view/diff.ex:199: Phoenix.LiveView.Diff.write_component/4
    (phoenix_live_view 0.16.4) lib/phoenix_live_view/channel.ex:483: Phoenix.LiveView.Channel.component_handle_event/6
    (stdlib 4.0.1) gen_server.erl:1120: :gen_server.try_dispatch/4
    (stdlib 4.0.1) gen_server.erl:1197: :gen_server.handle_msg/6
Last message: %Phoenix.Socket.Message{event: "event", join_ref: "4", payload: %{"cid" => 1, "event" => "start_actor_file_hotreload", "type" => "form", "value" => "_csrf_token=YS9dPRszDlsDfmAYU3UhMz87YB4vYQw_SIoUVFdjM9USgCFVOy6gk49m&path=%2Fpath%2Fto%2Fmyactor_s.wasm&count=1"}, ref: "8", topic: "lv:phx-Fzbamyqx_P2RtwAi"}
11:52:27.703 [info] Elixir.Gnat.ConsumerSupervisor starting graceful shutdown
11:52:28.204 [info] Elixir.Gnat.ConsumerSupervisor finished graceful shutdown
```
Demonstrating the site of the error by running `./bin/wasmcloud_host remote` and:
```
iex> {:ok, backend} = FileSystem.Backend.backend(:fs_mac)
{:ok, FileSystem.Backends.FSMac}
iex> args = [worker_pid: self(), dirs: ["/path/to/myactor_s.wasm"]]
[
  worker_pid: #PID<0.2308.0>,
  dirs: ["/path/to/myactor_s.wasm"]
]
iex> memo = backend.start_link(args)
** (EXIT from #PID<0.2308.0>) shell process exited with reason: an exception was raised:
    ** (ErlangError) Erlang error: :eacces
        :erlang.open_port({:spawn_executable, '/Users/me/.wash/downloads/lib/file_system-0.2.10/priv/mac_listener'}, [:stream, :exit_status, {:line, 16384}, {:args, ['-F', '/Users/me/Desktop/src/personal/verbalcode/build/verbalcode_s.wasm']}, {:cd, "/var/folders/r5/0kbwx40j3r55w4gbc2klj4300000gn/T/"}])
        (file_system 0.2.10) lib/file_system/backends/fs_mac.ex:143: FileSystem.Backends.FSMac.init/1
        (stdlib 4.0.1) gen_server.erl:848: :gen_server.init_it/2
        (stdlib 4.0.1) gen_server.erl:811: :gen_server.init_it/6
        (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```
Which is fixed by manually changing permissions: `chmod a+x lib/file_system-0.2.10/priv/mac_listener`

brooksmtownsend reviewed (commented):
Amazing catch here! Thanks for contributing this, I'll run the CI and give it a shot on my local mac but this looks good from a first glance 🚀

brooksmtownsend commented:
@nihil2501 you will just have to take a look at the DCO check because we require signed commits in the wasmCloud organization. If you click **Details** by the failed check it will let you know the commands you need to run

nihil2501 commented:
> Amazing catch here! Thanks for contributing this, I'll run the CI and give it a shot on my local mac but this looks good from a first glance 🚀

Thanks. I didn't take the time to do a full manual test where I verify that this change to `wash` actually fixes my system. I only got as far as composing that test example and seeing that permissions changed before and after this diff. Also I was doubtful that the exact form of my changes would be appropriate to merge. It only addresses the issue for `macOS` and also the test example could potentially be less ad-hoc.

brooksmtownsend commented:
> Thanks. I didn't take the time to do a full manual test where I verify that this change to `wash` actually fixes my system. I only got as far as composing that test example and seeing that permissions changed before and after this diff. Also I was doubtful that the exact form of my changes would be appropriate to merge. It only addresses the issue for `macOS` and also the test example could potentially be less ad-hoc.

Including a test was a welcome addition. The `mac_listener` is, maybe not surprisingly, a Mac only thing and each OS approaches file watching a little differently so this is an acceptable thing to put in to fix Mac only.

nihil2501 commented:
> > Thanks. I didn't take the time to do a full manual test where I verify that this change to `wash` actually fixes my system. I only got as far as composing that test example and seeing that permissions changed before and after this diff. Also I was doubtful that the exact form of my changes would be appropriate to merge. It only addresses the issue for `macOS` and also the test example could potentially be less ad-hoc.
> 
> Including a test was a welcome addition. The `mac_listener` is, maybe not surprisingly, a Mac only thing and each OS approaches file watching a little differently so this is an acceptable thing to put in to fix Mac only.

It does seem like the way `filesystem` handles `macos` specifically needs this treatment.
`macos`: https://github.com/falood/file_system/blob/main/mix.exs#L54-L55
other `unix`: https://github.com/falood/file_system/blob/main/lib/file_system/backends/fs_inotify.ex#L71

So I'll see if I can run this version of `wash` locally and manually verify the fix. Also, I did consolidate that ad-hoc test into `can_download_wasmcloud_tarball`. Then if that all that works and PR is approved I'll sign the DCO and merge.

nihil2501 commented:
> > Thanks. I didn't take the time to do a full manual test where I verify that this change to `wash` actually fixes my system. I only got as far as composing that test example and seeing that permissions changed before and after this diff. Also I was doubtful that the exact form of my changes would be appropriate to merge. It only addresses the issue for `macOS` and also the test example could potentially be less ad-hoc.
> 
> Including a test was a welcome addition. The `mac_listener` is, maybe not surprisingly, a Mac only thing and each OS approaches file watching a little differently so this is an acceptable thing to put in to fix Mac only.

@brooksmtownsend It worked locally after all.

nihil2501 commented on crates/wash-lib/src/start/wasmcloud.rs:
```diff
@@ -336,8 +337,8 @@ mod test {
     use reqwest::StatusCode;
-    use std::{collections::HashMap, env::temp_dir};
-    use tokio::fs::{create_dir_all, remove_dir_all};
+    use std::{collections::HashMap, env::temp_dir, os::unix::prelude::PermissionsExt};
```
@brooksmtownsend Good catch--fixed. I guess CI isn't exercised until after PR approval? Would that defect have shown up in CI?

thomastaylor312 commented on crates/wash-lib/src/start/wasmcloud.rs:
```diff
@@ -336,8 +337,10 @@ mod test {
+    use std::os::unix::prelude::PermissionsExt;
     use std::{collections::HashMap, env::temp_dir};
-    use tokio::fs::{create_dir_all, remove_dir_all};
+    use tokio::fs::{create_dir_all, remove_dir_all, File};
```
It might be easier to add this `File` and `std::os::unix::prelude::PermissionsExt` imports into the match block below. It would make things a little cleaner and avoid the "unused import `File`" warning on non-unix systems

nihil2501 replied:
@thomastaylor312 I moved the conditional `use` to right above the test that uses it. Is this what you meant by "match block below"? Also I got rid of that direct import of `File` (fully qualifying at the call site instead). Somehow I didn't notice that that happened.

thomastaylor312 replied:
Yes that is correct, thanks for reading into what I said 😅.

thomastaylor312 commented on crates/wash-lib/src/start/wasmcloud.rs:
```diff
@@ -365,6 +368,31 @@ mod test {
+            let mut fs_dir: Option<std::path::PathBuf> = None;
+            let lib_dir = download_dir.join("lib");
+
+            for entry in std::fs::read_dir(lib_dir).unwrap() {
```
Super nit: You are running blocking IO in an async method, but this is a test, so there is no need to change it if you don't want to

nihil2501 replied:
Thanks, I went the async route.

nihil2501 commented:
Don't mean to take too much time, but does anyone have a suggestion for git workflow here? I'm thinking I can keep commits addressing PR feedback separate and sign it as required by DCO even though it's only temporarily necessary. And then once the review process is over I'd squash down?

brooksmtownsend commented:
@nihil2501 One thing you can do is simply add the `-s` or I think `--sign-off` to your `git commit` command, that automatically signs your commits so DCO won't complain every time. 

FWIW, what I do each time now is:
```
git add file1 file2
git commit -sm "my message"
git push blah blah
```

nihil2501 commented:
@brooksmtownsend @thomastaylor312 I think this is all ready to go if one of you would like to sign off.

brooksmtownsend reviewed (approved):
I think this looks great as long as @thomastaylor312 had his questions addressed

nihil2501 commented:
> I think this looks great as long as @thomastaylor312 had his questions addressed

Thanks all. That may have been my first OSS contribution 👍
